### PR TITLE
fix(cli): distinguish between subjectSet and subjectID in check command

### PR DESCRIPTION
**Pull request by treilik**
attempt of fixing #850 but the command:

`./keto check 'videos:/cats/1.mp4#owner' view videos '/cats/1.mp4'`
still Outputs "Denied" even though it should be "Allowed" because of:
```
NAMESPACE       OBJECT          RELATION NAME   SUBJECT
videos          /cats/1.mp4     view            videos:/cats/1.mp4#owner

```

### CODE REVIEW

File: cmd/check/root.go [hunk 1]
```diff
@@ -42,13 +44,35 @@ func newCheckCmd() *cobra.Command {
...
 
+			subject := args[0]
+			relation := args[1]
+			namespace := args[2]
+			object := args[3]
+
+			var s *rts.Subject
+
+			// distinguish between subjectSet and subjectID
+			if strings.Contains(subject, "#") {
```
Comment (hperl) [inline cmd/check/root.go:55]: I'd rather distinguish based on `:` here, since the relation could be empty (= wildcard) for a subject set, e.g., `check users:Foo access files README.md`.

**Comment by hperl** [comment]
Hi, and thanks for your contribution! It looks quite solid already, but could you add some tests to see that it actually works? Thanks!

**Comment by treilik** [comment]
> Hi, and thanks for your contribution! It looks quite solid already, but could you add some tests to see that it actually works? Thanks!

Hi, Yes I'll do :)

File: ketoapi/enc_proto.go [hunk 2]
```diff
@@ -57,6 +57,15 @@ func (r *RelationTuple) ToProto() *rts.RelationTuple {
 	return res
 }
 
+func (s *SubjectSet) ToProto() *rts.SubjectSet {
```
Comment (zepatrik) [inline ketoapi/enc_proto.go:60]: It seems this is not used? I'd recommend to remove it again then :wink:

File: internal/e2e/cases_test.go [hunk 3]
```diff
@@ -28,6 +28,38 @@ func runCases(c client, m *namespaceTestManager) func(*testing.T) {
...
+				Relation:  rel,
+			}
+			rt := &ketoapi.RelationTuple{
+				Object:     obj,
+				SubjectSet: ss,
+				Namespace:  n.Name,
+				Relation:   rel,
+			}
+
+			assert.True(t, c.check(t, rt))
```
Comment (zepatrik) [inline internal/e2e/cases_test.go:61]: The reason why this test is failing is that the created tuples are not related to the checked tuple. It would be sufficient here to just create one tuple with a subject set and then check for the same tuple.

File: cmd/check/root.go [hunk 4]
```diff
@@ -42,13 +44,35 @@ func newCheckCmd() *cobra.Command {
...
+				su := &ketoapi.SubjectSet{}
+				su, err := su.FromString(subject)
+				if err != nil {
+					return err
+				}
+
+				s = rts.NewSubjectSet(su.Namespace, su.Object, su.Relation)
+			} else {
+				s = rts.NewSubjectID(subject)
+			}
```
Comment (zepatrik) [inline cmd/check/root.go:65]: I'd prefer to make this a small helper function that has its own unit test. That way we can specifically test for the subject parsing of the CLI with more edge cases. The e2e test makes sense as it is right now as well just to ensure that subject sets are parsed and supported by all clients.

**Review by zepatrik — CHANGES_REQUESTED** [review]
Thanks, sorry for the late review. Had a few out-of-office days.
There is just a few small things needed for the test to pass :wink:
